### ISSUE TITLE
Use of transformers patch

### ISSUE THREAD

**Issue by clefourrier** [IBM/transformers-struct-guidance#1]
Hi!
I do not understand what the transformer patch is for. 

If I understand it properly, it seems to me that it adds a check for on the size of the attention_mask, to skip an expansion step if its size already is 4 - though expanding a tensor of already the right size should not raise a problem in theory. 

So, was this added to gain computation time? Is it a memory problem? What is it for? 

Thank you for your time!

**Comment by pqian11** [comment]
Hi, thanks for your question! The patch is used to support structured attention mask, as illustrated in Section 2.2 in the [paper](https://arxiv.org/pdf/2108.00104.pdf). You can take a look at the code [here](https://github.com/IBM/transformers-struct-guidance/blob/dda3f92829b2e3bb86fb28fc7d1a8fa5f6ae7af5/src/plm-gen.py#L590) and [here](https://github.com/IBM/transformers-struct-guidance/blob/dda3f92829b2e3bb86fb28fc7d1a8fa5f6ae7af5/src/plm-gen.py#L432) to learn about how the structured attention mask is implemented. Basically for certain attention heads, some tokens in the input context are masked based on the parser state at each time step. With the patch applied, we can use the customized attention mask tensor as input to the model function.
